src/cdk/stepper/stepper.ts

```diff
@@ -0,0 +1,178 @@
+  oldStep: CdkStep;
+}
+
+@Component({
+  selector: 'cdk-step',
+  templateUrl: 'step.html',
+})
+export class CdkStep {
+  /** Template for step label if it exists. */
+  @ContentChild(CdkStepLabel) stepLabel: CdkStepLabel;
+
+  /** Template for step content. */
+  @ViewChild(TemplateRef) content: TemplateRef<any>;
+
+  /** Label of the step. */
+  @Input()
+  label: string;
+
+  constructor(private _stepper: CdkStepper) { }
+
+  /** Selects this step component. */
+  select(): void {
+    this._stepper.select(this);
+  }
+}
+
+@Directive({
+  selector: 'cdkStepper',
+  host: {
+    '(focus)': '_setStepfocused()',
+    '(keydown)': '_onKeydown($event)',
+  },
+})
+export class CdkStepper {
+  /** The list of step components that the stepper is holding. */
+  @ContentChildren(CdkStep) _steps: QueryList<CdkStep>;
+
+  /** The list of step headers of the steps in the stepper. */
+  @ViewChildren('stepHeader') _stepHeader: QueryList<ElementRef>;
+
+  /** The index of the currently selected step. */
+  @Input()
+  get selectedIndex() { return this._selectedIndex; }
+  set selectedIndex(value: any) {
+    this._selectedIndex = coerceNumberProperty(value);
+  }
+  private _selectedIndex: number;
+
+  /** Event emitted when the selected step has changed. */
+  @Output() selectionChange = new EventEmitter<CdkStepperSelectionEvent>();
+
+  /** The index of the step that the focus can be set. */
+  _focusIndex: number = 0;
+
+  /** Used to track unique ID for each stepper component. */
+  private _groupId: number;
+
+  constructor() {
+    this._groupId = nextId++;
+  }
+
+  /** Selects and focuses the provided step. */
+  select(step: CdkStep | number): void {
+    if (typeof step == 'number') {
+      this._emitStepperSelectionEvent(step, this._selectedIndex);
+    } else {
+      let stepsArray = this._steps.toArray();
+      this._emitStepperSelectionEvent(stepsArray.indexOf(step), this._selectedIndex);
+    }
+    this._setStepFocused(this._selectedIndex);
+  }
+
+  /** Selects and focuses the next step in list. */
+  next(): void {
+    if (this._selectedIndex == this._steps.length - 1) { return; }
+    this._emitStepperSelectionEvent(this._selectedIndex + 1, this._selectedIndex);
+    this._setStepFocused(this._selectedIndex);
+  }
+
+  /** Selects and focuses the previous step in list. */
+  previous(): void {
+    if (this._selectedIndex == 0) { return; }
+    this._emitStepperSelectionEvent(this._selectedIndex - 1, this._selectedIndex);
+    this._setStepFocused(this._selectedIndex);
+  }
+
+  /** Returns a unique id for each step label element. */
+  _getStepLabelId(i: number): string {
+    return `mat-step-label-${this._groupId}-${i}`;
+  }
+
+  /** Returns nique id for each step content element. */
+  _getStepContentId(i: number): string {
+    return `mat-step-content-${this._groupId}-${i}`;
+  }
+
+  private _emitStepperSelectionEvent(newIndex: number,
+                                       oldIndex: number): void {
```

Comment:
is `oldIndex` not just always whatever `this._selectedIndex` at the start of this method?